### PR TITLE
feat(api): NetworkPort to provide direct connections

### PR DESCRIPTION
**Pull request by po1vo**
`with_connections` option now applies to NetworkPort.

I'm just not sure what the output should be:
`$fields['_connections']['NetworkPort'] = $otherPortID;`
or
`$fields['_connections']['NetworkPort'] = [$otherPortID];`
or
`$fields['_connections'] = $otherPortID;`
?

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A

### CODE REVIEW

File: src/Api/API.php [hunk 1]
```diff
@@ -737,41 +737,48 @@ protected function getItem($itemtype, $id, $params = [])
...
+                            $fields['_connections'][$connect_type][] = $data;
+                        }
                     }
                 }
+            } else if ($itemtype == "NetworkPort") {
+                $NetworkportNetworkport = new \NetworkPort_NetworkPort();
+                $otherPortID = $NetworkportNetworkport->getOppositeContact($id);
+                if ($otherPortID) {
+                    $fields['_connections']['NetworkPort'] = $otherPortID;
+                }
```
Comment (stonebuzz) [inline src/Api/API.php:781]:
```suggestion
         $networkport_networkport = new \NetworkPort_NetworkPort();
         $opposite_networkport_id = $networkport_networkport->getOppositeContact($id);
         if ($opposite_networkport_id) {
            $fields['_connections']['NetworkPort'][] = [
               "assoc_opposite_networkports_id" => $opposite_networkport_id,
               "assoc_networkports_id" => $id,
               "assoc_id" => $networkport_networkport->fields['id']
            ]
         }
```

Comment (stonebuzz) [reply src/Api/API.php:781]: Add all info from assoc object (```Networkport_Networkport```)

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Api/API.php (hunk 1 above)
Comment (po1vo) [reply src/Api/API.php:781]: All info from assoc object seems really excessive. A user would already have assoc_networkports_id just to run the request, and assoc_id is a mere table index. Do you insist on this change?

Comment (po1vo) [reply src/Api/API.php:781]: Moreover, we're returning a NetworkPort, not Networkport_Networkport, so we should just do:
```php
$fields['_connections']['NetworkPort'][] = ["assoc_id" => $opposite_networkport_id];
```

File: src/Api/API.php [hunk 2]
```diff
@@ -738,41 +738,51 @@ protected function getItem($itemtype, $id, $params = [])
...
                 }
+            } else if ($itemtype == 'NetworkPort') {
+                $networkport_networkport = new \NetworkPort_NetworkPort();
+                $opposite_networkport_id = $networkport_networkport->getOppositeContact($id);
+                if ($opposite_networkport_id) {
+                    $fields['_connections']['NetworkPort'][] = [
+                        'id'       => $opposite_networkport_id,
+                        'assoc_id' => $networkport_networkport->fields['id']
+                    ];
+                }
```
Comment (stonebuzz) [inline src/Api/API.php:785]: So wee need this
```suggestion
            } else if ($itemtype == 'NetworkPort') {
         $networkport_networkport = new \NetworkPort_NetworkPort();
         $opposite_networkport_id = $networkport_networkport->getOppositeContact($id);
         if ($opposite_networkport_id) {
            $fields['_connections']['NetworkPort'][] = [
               "assoc_networkports_id_2" => $opposite_networkport_id,
               "assoc_networkports_id_1" => $id,
               "assoc_id" => $networkport_networkport->fields['id']
            ]
         }
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Api/API.php (hunk 2 above)
Comment (po1vo) [reply src/Api/API.php:785]: Which key is the sought NetworkPort `id` in this output? Do you suggest a user compare his input id to both `assoc_networkports_id_1` and `assoc_networkports_id_2` to find the opposite port id?

Moreover the provided output doesn't reflect the true state of the DB: `$id` may be either `assoc_networkports_id_1` or `assoc_networkports_id_2`.

In summary, there is no `NetworkPort` in the output and sorry, but i this change makes no sense to me.

**Comment by cedric-anne** [comment]
Hi,

Sorry for the really late response. We will not accept this PR as it will produce a totally different behaviour between computers and network ports. The `with_connections` parameter is used to indicates that the response should returned items that have a direct connection, i.e. a physical connection that make them part of the workstation. It has nothing to do with network connections.

I do not know enough the API to tell you were is the best place to expose opposite device connected through networkports. Maybe by adding information in data returned by `with_networkports`, or by adding a new parameter. @orthagh Have you an idea about this?

**Comment by po1vo** [comment]
> The `with_connections` parameter is used to indicates that the response should returned items that have a direct connection, i.e. a physical connection that make them part of the workstation. It has nothing to do with network connections.

It looks like you overlooked that the patch only enhances NetworkPorts. And in case of NetworkPorts, what other kind connections can there be if not network?

> I do not know enough the API to tell you were is the best place to expose opposite device connected through networkports.

It seems pretty logical to return network connections when requesting `/NetworkPort/:id?with_connections`

**Comment by cedric-anne** [comment]
> It looks like you overlooked that the patch only enhances NetworkPorts.

No. My point is that a parameter should always act the same, when it does, whatever the itemtype used.

**Comment by cedric-anne** [comment]
@po1vo 

We have discussed this point internally and we all agree that reusing an existing parameter name and assigning it a different behavior is a bad idea. It may generate confusion.

So, the solution is to either add some more information when `with_networkports` parameter is used, either adding a new parameter to get this information.

**Comment by po1vo** [comment]
> @po1vo
> 
> We have discussed this point internally and we all agree that reusing an existing parameter name and assigning it a different behavior is a bad idea. It may generate confusion.
> 
> So, the solution is to either add some more information when `with_networkports` parameter is used, either adding a new parameter to get this information.

I'll take the `with_networkports` route.

**Comment by cedric-anne** [comment]
:+1:

**Comment by po1vo** [comment]
I took a look at `Api::getNetworkPorts` (which is used by `with_networkports` param) code and saw that it returns `NetworkName`-s instead of `NetworkPort`-s, so reusing this parameter makes little sense.

I still believe that within `NetworkPort` context a connection is a connection to another network port, so `with_connections` parameter would make total sense, but i don't want to argue anymore, so I've decided to close the PR. Thank you for all of your input.

**Pull request closed without merging**